Implementing Py42ActiveLegalHoldError

DiscoRiver opened the pull request:
### Description of Change ###

Added `Py42ActiveLegalHoldError` exception. Raised when attempting to deactivate user/device that is in an active legal hold. 

### Issues Resolved ###

- Fixes #169 

### Testing Procedure ###

1. Use `py42.sdk.devices.deactivate()` or `py42.sdk.users.deactivate()` to deactivate a user in an active legal hold.
2. Observe message `Cannot deactivate from user in an active legal hold.`

### Notes ###

- Create draft PR as I suspect messaging for this exception will need to be updated, and I want to clarify the approach I've taken with this is acceptable before it's reviewed proper.

DiscoRiver commented on src/py42/exceptions.py:
```diff
@@ -107,6 +109,15 @@ class Py42InternalServerError(Py42HTTPError):
+    active legal hold."""
+
+    def __init__(self, exception):
+        msg = u"Cannot deactivate from user in an active legal hold."
```
I suspect we may want to update this message.

unparalleled-js replied:
Could add a `resource` parameter that is either `user` or `device`, along with an `uid` parameter that is either the `device_guid` or the `userUid` and the message could be like `u"Cannot deactivate the {0} with UID {1} as the {0} is involed in a legal hold matter.".format(resource, uid)`.

DiscoRiver commented on src/py42/exceptions.py:
```diff
@@ -153,7 +164,13 @@ def raise_py42_error(raised_error):
     """
     if raised_error.response.status_code == 400:
-        raise Py42BadRequestError(raised_error)
+        if json.loads(raised_error.response.text)[0]["name"] in [
```
Since the response text comes back as a string, and the code is simply `400`, I didn't see a better approach than this. We want to capture it before a `Py42BadRequestError` is raised.

unparalleled-js replied:
Could we not just do what we are doing in `py42.services.detectionlists.__init__.py` merthod: `handle_user_already_added_errpr` ?

I may not understand the problem, but see that method if you haven't already.  It checks the bad request error (a py42 error) for the error text.

DiscoRiver replied:
That's a much better way of doing it! I've updated based on this design.

unparalleled-js commented on src/py42/exceptions.py:
```diff
@@ -153,7 +164,13 @@ def raise_py42_error(raised_error):
-        raise Py42BadRequestError(raised_error)
+        if json.loads(raised_error.response.text)[0]["name"] in [
+            "USER_IN_ACTIVE_LEGAL_HOLD",
+            "ACTIVE_LEGAL_HOLD",
```
Could just `json.loads(raised_error.response.text)[0][u"name"].contains(u"ACTIVE_LEGAL_HOLD")` to encompass both

DiscoRiver replied:
I've moved all of this out of `py42.exceptions` based on your other comments. Looking back, it added overhead to any `400` response, which wasn't ideal.

DiscoRiver commented:
I changed this from a draft because I think it's pretty much complete.

unparalleled-js commented on CHANGELOG.md:
```diff
@@ -16,6 +16,8 @@ how a consumer would use the library (e.g. adding unit tests, updating documenta
 
 ### Added
 
+- Added `Py42ActiveLegalHoldError` exception when attempting to deactivate a user or device in an active legal hold.
```
Not a big deal here because it is obvious, but for SDK type projects, it is better to mention the exact methods being affected. In this case it would be `sdk.users.deactivate()` and `sdk.devices.deactivate()`.  Just 1 less mental connection we have to put on CHANGELOG consumers.

unparalleled-js reviewed (changes requested):
I approved too soon! We need tests :)

Both `users.deactivate()` and `devices.deactivate()` should be tested to make sure that when the connection raises x error containing y text, than it raises this new error ` Py42ActiveLegalHoldError `

DiscoRiver commented:
> I approved too soon! We need tests :)
> 
> Both `users.deactivate()` and `devices.deactivate()` should be tested to make sure that when the connection raises x error containing y text, than it raises this new error `Py42ActiveLegalHoldError`

My bad. Could've sworn I added those. Just updated :)
